packages/net-stubbing/lib/server/state.ts

```diff
@@ -1,24 +1,60 @@
-import { noop } from 'lodash'
+import type { InterceptedRequest } from './intercepted-request'
 import type { NetStubbingState } from './types'
 
 export function state (): NetStubbingState {
   return {
     requests: {},
     routes: [],
     pendingEventHandlers: {},
-    reset () {
-      // clean up requests that are still pending
-      for (const requestId in this.requests) {
-        const { res } = this.requests[requestId]
+    async reset () {
+      await Promise.all(Object.values <InterceptedRequest>(this.requests).map((request) => {
+        const responseEvents = ['before:response', 'response:callback', 'response', 'network:error']
+        const inResponsePhase = responseEvents.includes(request.lastEvent!)
+        const { res } = request
 
-        res.removeAllListeners('finish')
-        res.removeAllListeners('error')
-        res.on('error', noop)
-        res.destroy()
-      }
+        if (res.destroyed) {
+          delete this.requests[request.id]
+
+          return Promise.resolve()
+        }
+
+        let shouldDestroyResponse = false
+
+        request.subscriptionsByRoute.forEach((subscriptionByRoute) => {
+          if (!subscriptionByRoute.immediateStaticResponse) {
+            subscriptionByRoute.subscriptions.forEach((subscription) => {
+              if (subscription.await && !subscription.skip && (responseEvents.includes(subscription.eventName) || !inResponsePhase)) {
+                shouldDestroyResponse = true
+              }
+            })
+          }
+        })
+
+        if (shouldDestroyResponse) {
+          res.removeAllListeners('finish')
+          res.removeAllListeners('error')
+
+          return new Promise<void>((resolve) => {
+            res.once('finish', () => {
+              res.destroy()
+              delete this.requests[request.id]
+              resolve()
+            })
+
+            res.once('error', () => {
+              res.destroy()
+              delete this.requests[request.id]
+              resolve()
+            })
+
+            res.end()
+          })
+        }
+
+        return Promise.resolve()
+      }))
 
       this.pendingEventHandlers = {}
-      this.requests = {}
```

Comment:
I see you've switched from using `this.requests = {}` to using `delete this.requests[requestId]`, but is there a reason to change this? is existing code referencing `this.requests[requestId]` and causing errors?

I'd argue that we don't want to get rid of this line - `InterceptedRequest` objects contain references to `BackendRoutes`, and we're trying to prevent leaking these route objects between tests to eliminate race conditions, so it seems like this object should be totally empty between tests, right?